cmd/cvc-operator/controller/cstorvolumeclaim.go

```diff
@@ -643,7 +668,376 @@ func addPDBLabelOnCVC(
+// ReplicaCount. If CVC holds any volume replica pool information then verify
+// with Status.PoolInfo
 func isHAVolume(cvcObj *apis.CStorVolumeClaim) bool {
-	return cvcObj.Spec.ReplicaCount >= minHAReplicaCount
+	if len(cvcObj.Status.PoolInfo) == 0 {
+		return cvcObj.Spec.ReplicaCount >= minHAReplicaCount
+	}
+	return len(cvcObj.Status.PoolInfo) >= minHAReplicaCount
+}
+
+// 1. If Volume was already pointing to a PDB then check is that same PDB will be
+//    applicable after scalingup/scalingdown(case might be from 4 to 3
+//    replicas) if applicable then return same pdb name. If not applicable do
+//    following changes:
+//    1.1 Delete PDB if no other CVC is pointing to PDB.
+// 2. If current volume was not pointing to any PDB then do nothing.
+// 3. If current volume is HAVolume then check is there any PDB already
+//    existing among the current replica pools. If PDB exists then return
+//    that PDB name. If PDB doesn't exist then create new PDB and return newely
+//    created PDB name.
+// 4. If current volume is not HAVolume then return nothing.
+func updatePDBForVolume(cvcObj *apis.CStorVolumeClaim) (string, error) {
+	pdbName, hasPDB := cvcObj.GetLabels()[string(apis.PodDisruptionBudgetKey)]
+	pdbLabels := cvclaim.GetPDBPoolLabels(cvcObj.Status.PoolInfo)
+	labelSelector := apispdb.GetPDBLabelSelector(pdbLabels)
+	if hasPDB {
+		// Get PDB if exists among newely updated volume replicas pools
+		pdbList, err := apispdb.KubeClient().
+			WithNamespace(openebsNamespace).
+			List(metav1.ListOptions{LabelSelector: labelSelector})
+		if err != nil {
+			return "", errors.Wrapf(err,
+				"failed to get PDB present among pools %v",
+				cvcObj.Status.PoolInfo,
+			)
+		}
+		if len(pdbList.Items) >= 1 && isHAVolume(cvcObj) {
+			for _, pdbObj := range pdbList.Items {
+				pdbPoolCount := len(pdbObj.Spec.Selector.MatchExpressions[0].Values)
+				// Let us assume that volume replicas was scale down from 4 to
+				// 3(i.e PDB was created on top of 4 pools). Now when scale down
+				// happens better to delete the PDB(if no one refering to it) and
+				// create PDB among 3 pools so that scaling down of cluster will
+				// not hold unnecessarily(i.e draining the node).
+				if pdbObj.Name == pdbName && pdbPoolCount < len(cvcObj.Status.PoolInfo) {
+					return pdbName, nil
+				}
+			}
+		}
+		err = deletePDBIfNotInUse(cvcObj)
+		if err != nil {
+			return "", err
+		}
+	}
+	if !isHAVolume(cvcObj) {
+		return "", nil
+	}
+	pdbObj, err := getOrCreatePodDisruptionBudget(getCSPC(cvcObj), cvcObj.Status.PoolInfo)
+	if err != nil {
+		return "", err
+	}
+	return pdbObj.Name, nil
+}
+
+// isCVCScalePending returns true if there is change in desired replica pool
+// names and current replica pool names
+// 1. Below function will check whether there is any change in desired replica
+//    pool names and current replica pool names.
+func (c *CVCController) isCVCScalePending(cvc *apis.CStorVolumeClaim) bool {
```

Comment:
add a check that cvc is in bound state.. if not, return false

Reply:
and also add comment that scale up/down of cvc will not work until cvc is in bound state

Reply:
Not required to add a check. Resize, scale Up/Down will execute only once CVC is bound.